### ISSUE TITLE
Let's Talk: Constructor Injection

### ISSUE THREAD

**Issue by Boundarybreaker** [FabricMC/Mixin#26]
The new 1.15 snapshots move almost all of `MinecraftClient`'s setup stuff into the constructor. This is a *massive* hassle for any mod that wants to add its own resource pack stuff, which includes Fabric API. Fabric API has [resorted to a `@Redirect`](https://github.com/FabricMC/fabric/blob/1.15/fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/MixinMinecraftGame.java#L74), but this means that nobody else can use that point, since `@Redirect` is collidable. The reason we have this problem is because `@Inject` only allows injection to the `RETURN` of constructors. Do we want to open that up to any other injection points? The main worry is preventing anything from being put before a `super` call.

**Comment by liach** [comment]
fyi mixin has a check to allow certain types of injections to be done after super calls; it generally groups injections to performable before super, after super, only in regular method body (these 3 types)

**Comment by Chocohead** [comment]
Injecting before super is called is perfectly possible if the callback is forced to be static. In fact it can be quite useful want wanting to modify a parameter passed up to the superclass as it avoids needing to catch every instance of the constructor being used. Not sure if Mixin properly supports this for all injection types though (similar to it allowing redirecting `<init>` which leaves uninitialised references about).

For what it's worth Rift always allowed you to tinker in constructors as you could in any other method and it never had problems with people doing stupid stuff. Mixin either lets you and is careful enough to avoid problems or crashes immediately (which is as much down to bugs within Mixin as it is safety features).

**Comment by Boundarybreaker** [comment]
yeah I'm super in favor of more opened init injections tbh, it'd make things a lot easier just in general

**Comment by liach** [comment]
@Boundarybreaker #30 has been made and works well. Feel free to peek [over there](https://github.com/FabricMC/Mixin/pull/30).

**Comment by Devan-Kerman** [comment]
> if you want to touch <clinit> go nuts
I haven't tried, because in my opinion there's not enough reason to stray from "you can redirect, etc. but you can only @inject at RETURN or you can make a custom injection point to circumvent that"
that imo is enough, because it forces you to think before you break the rule, but still makes the rule fundamentally breakable if you are determined to play with fire
- Mumfrey

the enforcement of a static mixin will need a PR though, idk if liach's thing already does it or not

**Comment by liach** [comment]
My pr only supports regular injects after super call, not static injects passed to super calls

**Comment by valoeghese** [comment]
did https://github.com/FabricMC/Mixin/commit/536170a94675b07e462cc46cdb6493586bdd6dcc fix this
